File: sdk/go/transports/mock.go
package transports

import "sync"

// MockTransport is an in-memory transport for testing.
type MockTransport struct {
	mu          sync.Mutex
	connected   bool
	SentPackets [][]byte
	recvCh      chan []byte
}

// NewMock creates a new MockTransport.
func NewMock() *MockTransport {
	return &MockTransport{
		recvCh: make(chan []byte, 64),
	}
}

func (m *MockTransport) Connect() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.connected = true
	return nil
}

func (m *MockTransport) Close() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.connected = false
	return nil
}

func (m *MockTransport) Send(packet []byte) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	cp := make([]byte, len(packet))
	copy(cp, packet)
	m.SentPackets = append(m.SentPackets, cp)
	return nil
}

func (m *MockTransport) Recv() ([]byte, error) {
	pkt := <-m.recvCh
	return pkt, nil
}

func (m *MockTransport) Connected() bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.connected
}

func (m *MockTransport) NeedsCOBS() bool { return false }

// Inject simulates a packet received from the device.
func (m *MockTransport) Inject(packet []byte) {
	m.recvCh <- packet
}

File: sdk/js/src/transports/mock.ts
/**
 * GRAFT Mock Transport
 *
 * In-memory transport for testing. Supports injecting responses
 * and capturing sent packets without any real I/O.
 */

import type { GraftTransport } from './transport.js'

export class MockTransport implements GraftTransport {
  private _connected = false
  private _handler: ((packet: Uint8Array) => void) | null = null
  private _sentPackets: Uint8Array[] = []

  readonly needsCOBS = false

  get connected(): boolean {
    return this._connected
  }

  /** All packets sent by the SDK through this transport. */
  get sentPackets(): Uint8Array[] {
    return this._sentPackets
  }

  /** The most recently sent packet, or undefined. */
  get lastSent(): Uint8Array | undefined {
    return this._sentPackets[this._sentPackets.length - 1]
  }

  async connect(): Promise<void> {
    this._connected = true
  }

  async disconnect(): Promise<void> {
    this._connected = false
    this._handler = null
  }

  async send(packet: Uint8Array): Promise<void> {
    if (!this._connected) throw new Error('MockTransport: not connected')
    this._sentPackets.push(new Uint8Array(packet))
  }

  onReceive(handler: (packet: Uint8Array) => void): void {
    this._handler = handler
  }

  /**
   * Inject a packet as if the device sent it.
   * The SDK's receive handler will be called synchronously.
   */
  inject(packet: Uint8Array): void {
    if (this._handler) {
      this._handler(new Uint8Array(packet))
    }
  }

  /** Clear all recorded sent packets. */
  clearSent(): void {
    this._sentPackets.length = 0
  }

  /** Reset the transport to initial state. */
  reset(): void {
    this._connected = false
    this._handler = null
    this._sentPackets.length = 0
  }
}
